File: apps/server/src/persist/webhook.ts
export interface WebhookRanking {
  playerId: string;
  nome: string;
  pontos: number;
  posicao: number;
}

export interface WebhookTitles {
  kamikaze: string | null;
  balaPerdida: string | null;
  covardeEstrategico: string | null;
}

export interface WebhookPayload {
  roomId: string;
  finalizadaEm: string;
  ranking: WebhookRanking[];
  titulos: WebhookTitles;
}

/**
 * Dispara o ranking final da partida para o n8n postar no canal. Falha de webhook nunca derruba
 * a partida — qualquer erro (rede, DNS, timeout, 4xx/5xx) é só logado.
 */
export async function sendMatchWebhook(payload: WebhookPayload): Promise<void> {
  const url = process.env.WEBHOOK_URL;
  if (!url) return;

  try {
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      console.error(`[webhook] resposta não-ok do WEBHOOK_URL: ${response.status} ${response.statusText}`);
    }
  } catch (err) {
    console.error('[webhook] falha ao notificar WEBHOOK_URL:', err);
  }
}

File: go/persist/webhook.go
package persist

import (
	"bytes"
	"context"
	"encoding/json"
	"log"
	"net/http"
	"os"
	"time"
)

// Porte de `apps/server/src/persist/webhook.ts`: dispara o ranking final para o n8n postar no
// canal. Falha de webhook NUNCA derruba a partida — qualquer erro (rede, DNS, timeout, 4xx/5xx) é
// só registrado.

// EnviarWebhook posta o resumo da partida em `WEBHOOK_URL`. Sem a variável, não faz nada.
func EnviarWebhook(payload any) {
	url := os.Getenv("WEBHOOK_URL")
	if url == "" {
		return
	}
	corpo, err := json.Marshal(payload)
	if err != nil {
		log.Printf("[webhook] payload inválido: %v", err)
		return
	}

	ctx, cancelar := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancelar()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(corpo))
	if err != nil {
		log.Printf("[webhook] requisição inválida: %v", err)
		return
	}
	req.Header.Set("content-type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Printf("[webhook] falha ao notificar WEBHOOK_URL: %v", err)
		return
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		log.Printf("[webhook] resposta não-ok do WEBHOOK_URL: %d", resp.StatusCode)
	}
}
